File: algorithms/548. Split Array with Equal Sum/solution.py
class Solution(object):
    def splitArray(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        n = len(nums)
        sums = [0] * (n + 1)
        for i in xrange(n):
            sums[i + 1] = sums[i] + nums[i]
        for j in xrange(3, n - 3):
            s = set()
            for i in xrange(1, j - 1):
                if sums[i] == (sums[j] - sums[i+1]):
                    s.add(sums[i])
            for k in xrange(j + 2, n - 1):
                sjk = sums[k] - sums[j+1]
                skn = sums[n] - sums[k+1]
                if sjk == skn and sjk in s:
                    return True
        return False

File: algorithms/548. Split Array with Equal Sum/solution.cc
class Solution {
public:
    bool splitArray(vector<int>& nums) {
        vector<int> sum = { 0 };
        int n = nums.size();
        for ( int i = 0; i < n; i++ ) {
            sum.push_back(sum.back() + nums[i]);
        }
        int i, j, k;
        set<int> visited;
        for ( j = 3; j < n - 3; j++ ) {
            visited.clear();
            for ( i = 1; i < j - 1; i++ ) {
                if ( sum[i] == sum[j] - sum[i+1] ) {
                    visited.insert(sum[i]);
                }
            }
            for ( k = j + 2; k < n - 1; k++ ) {
                int s1 = sum[k] - sum[j+1];
                int s2 = sum[n] - sum[k+1];
                if ( s1 == s2 && visited.find(s1) != visited.end() ) {
                    return true;
                }
            }
        }
        return false;
    }
};
